### PR TITLE
Yul IR: Change evaluation order for operands of binary operators.

### LINKED ISSUE #11536: Expression evaluation order for binary operators in Sol->Yul codegen.
While the order of evaluation for expression is undefined and for that reason this is "fine" in principle, I'm wondering:

In legacy codegen, for binary expressions, we evaluate right-hand-side first, then left-hand-side (unless one is a literal).
In Yul codegen we do the opposite and evaluate left-hand-side first, then right-hand-side.

Is there any good reason for doing it like that in Yul codegen?

If there is not there are two reasons for switching to the same behaviour as legacy code generation:
- If there is no good reason for using diverging behaviour, the divergence is basically intentionally screwing people without reason (even though it'd only happen for "bad" code relying on undefined behaviour, but still).
- Evaluating the right-hand-side first allows for more-optimal stack layouts (since the right-hand-side will end up as right-most argument to a builtin call and has to be deeper in stack than the left-hand-side - while some of the operations are commutative, not all of them are, and even for commutative ones, it's additional stress for the code generation to exploit that).

Unless there is a reason for keeping the current behaviour, I vote for changing the order to match legacy code generation.

### PR DESCRIPTION
**Pull request by ekpyron**
Rationale in https://github.com/ethereum/solidity/issues/11536.
Fixes https://github.com/ethereum/solidity/issues/11536.
References https://github.com/ethereum/solidity/issues/11414.
References https://github.com/ethereum/solidity/issues/11535.

If there is any reason for not doing this I'm not aware of, feel free to close.

### CODE REVIEW

**Comment by ekpyron** [comment]
Actually, even the current evm code transform seems to profit from this gas-wise.
<details><summary>Click for a table of gas differences</summary>

| File name                                                                |   IR-optimized (%) |   Legacy-Optimized (%) |   Legacy (%) |
|--------------------------------------------------------------------------|--------------------|------------------------|--------------|
| externalContracts/snark.sol                                              |         0.0287153  |                0       |            0 |
| externalContracts/strings.sol                                            |        -0.0502829  |               -2.88971 |            0 |
| abiEncoderV2/abi_encode_calldata_slice.sol                               |        -0.90104    |                0       |            0 |
| abiEncoderV2/calldata_array.sol                                          |        -5.77189    |                0       |            0 |
| abiEncoderV1/abi_encode_calldata_slice.sol                               |        -0.90104    |                0       |            0 |
| viaYul/unary_operations.sol                                              |         0          |                0       |            0 |
| viaYul/array_memory_index_access.sol                                     |        -0.761844   |                0       |            0 |
| viaYul/array_storage_index_access.sol                                    |        -0.0170482  |                0       |            0 |
| array/byte_array_transitional_2.sol                                      |        -0.298078   |                0       |            0 |
| array/push/byte_array_push_transition.sol                                |        -0.159357   |                0       |            0 |
| array/copying/array_copy_storage_storage_different_base_nested.sol       |        -0.0203275  |                0       |            0 |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol             |        -0.00109976 |                0       |            0 |
| array/copying/array_nested_calldata_to_storage.sol                       |        -0.00444493 |                0       |            0 |
| array/copying/memory_dyn_2d_bytes_to_storage.sol                         |        -0.30028    |                0       |            0 |
| array/copying/array_copy_target_leftover.sol                             |        -0.0682701  |                0       |            0 |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol   |        -0.0459268  |                0       |            0 |
| array/copying/array_copy_storage_storage_different_base.sol              |        -0.0732999  |                0       |            0 |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol |        -0.00720985 |                0       |            0 |
| array/copying/bytes_storage_to_storage.sol                               |        -2.18538    |                0       |            0 |
| array/pop/byte_array_pop_long_storage_empty.sol                          |        -1.32669    |                0       |            0 |
| structs/struct_copy_via_local.sol                                        |        -0.0117943  |                0       |            0 |
| various/destructuring_assignment.sol                                     |        -0.018616   |                0       |            0 |
</details>

Granted, though, the real-world external contract tests seem hardly affected.

File: test/libsolidity/semanticTests/externalContracts/strings.sol [hunk 1]
```diff
@@ -51,9 +51,9 @@ contract test {
 // compileViaYul: also
 // ----
 // constructor()
-// gas irOptimized: 912777
+// gas irOptimized: 910871
 // gas legacy: 1188228
-// gas legacyOptimized: 771634
+// gas legacyOptimized: 749336
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/externalContracts/strings.sol:56]: How did this change?

Comment (hrkrshnn) [reply test/libsolidity/semanticTests/externalContracts/strings.sol:56]: Ah, because of abi code?

Comment (ekpyron) [reply test/libsolidity/semanticTests/externalContracts/strings.sol:56]: Hm... but that doesn't go through the IR code generator, does it? Without thinking about it too much, this seems rather weird to me, too...

Comment (hrkrshnn) [reply test/libsolidity/semanticTests/externalContracts/strings.sol:56]: I ran the test setting a breakpoint on the change and `compileViaYul: false`. It doesn't enter there. So this is indeed weird.

**Comment by ekpyron** [comment]
Since after the discussion of this and realizing that function argument evaluation order is left-to-right actually, we'll probably not merge this, just as a note:
In order to avoid false positives during fuzzing we could use a patched version of old code generation for fuzzing, which has https://github.com/ethereum/solidity/blob/e7bf1cc7e7eb99c6143da6a6aaa6b890b99fc3e8/libsolidity/codegen/ExpressionCompiler.cpp#L536 swapped with the next line.

(Note that the case where ``swap`` is true a few lines further up should only be cases, in which the order will not affect the outcome, so it's only the ``swap == false`` case that would need a patch)

File: test/libsolidity/semanticTests/externalContracts/strings.sol (hunk 1 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/externalContracts/strings.sol:56]: Might be worth figuring out what's going on there, then... my guess would be that it's either a bug in how ``isoltest`` calculates gas costs or some weird bug, s.t. whatever happened before the ``legacyOptimized`` run changes its behaviour (e.g. some change in inlining order in the abi code due to some state bleeding through or AST IDs changing - respectively *something* like that - I don't really see what kind of thing along those lines could *actually* be happening here :-)).

**Comment by leonardoalt** [comment]
@ekpyron should we still review this PR? Or has it been decided that won't be done?

**Comment by hrkrshnn** [comment]
@leonardoalt The comment https://github.com/ethereum/solidity/issues/11536#issuecomment-861634369 suggests that it's unlikely to get this merged. I'll keep this PR as draft for now.

I think @bshastry wanted to use this somehow in the fuzzer.

**Comment by bshastry** [comment]
JFYI: I've made a note of this change and am already using it for fuzzing!

**Comment by ekpyron** [comment]
I think it's fine to close this.

**Pull request closed without merging**